pandas/_libs/tslibs/conversion.pyx

```diff
@@ -220,9 +220,15 @@ cdef int64_t cast_from_unit(
             f"cannot convert input {ts} with the unit '{unit}'"
         ) from err
 
-    frac = ts - base
-    if p:
-        frac = round(frac, p)
+    if isinstance(ts, np.integer):
+        # If ts is an integer then the fractional component will always be
+        # zero. It helps to set this explicitly following changes to type
+        # promotion behavior in NEP 50 (GH 56996).
+        frac = 0
```

Comment:
I'm not sure if we can do this outright since pandas will still need to support numpy < 2 behavior. Maybe `ts` should be cast to `int` instead?

Reply:
Thanks for taking a look @mroeschke—I thought my fix would preserve NumPy < 2 behavior too, but maybe I am missing something subtle.  I switched to your suggested approach.  Let me know if that looks OK.